### ISSUE TITLE
discussion: replace waterline-sequel with a community-supported query builder

### ISSUE THREAD

**Issue by tjwebb** [balderdashy/waterline#605]
I'm wondering if the project would benefit from using a more popular and more thoroughly-maintained query builder such as [knexjs](http://knexjs.org/). Interested in hearing feedback.

**Comment by tjwebb** [comment]
hey @tgriesser I saw you made the speakers list for jsconf. Let's meet up and chat. sails.js is interested in using knex in a number of ways (namely, to replace [waterline-sequel](https://github.com/balderdashy/waterline-sequel) in a number of places), and it'd be great to get buy-in from both communities. (folks on our side of things are pretty enthusiastic :+1: )

**Comment by tgriesser** [comment]
Definitely! Yeah that'd be sweet - I'm in the process of streamlining a ton of the internals https://github.com/tgriesser/knex/pull/786 and beefing up transactions (automatic savepoints when transactions are nested, etc) and brainstorming a better approach for writing and extending adapters (so we can easily get mssql and built-in support some of the great new postgres features)

It'd be great to see where we can collaborate and not all simultaneously reinvent the wheel on a lot of this.

**Comment by dmarcelino** [comment]
I think a good approach for this would be to replace waterline-sequel for a "waterline-knex" with the same inputs and same (or better) outputs, that way we could use it as a drop in replacement in sails-postgresql and sails-mysql.

**Comment by mugli** [comment]
@tgriesser Just saw you jsconf video. Awesome talk! I saw you mentioning sails at the end of the talk. Could you and sails guys meet there and make any progress on integrating knex with waterline?

**Comment by gaurav21r** [comment]
@tjwebb @particlebanana @dmarcelino @tgriesser  Any Updates on this?

**Comment by dmarcelino** [comment]
Kind of... @tjwebb has been working on a version of sails-postgresql using knex. So far it passes all integration tests.

**Comment by sailsbot** [comment]
Thanks for posting, @tjwebb.  I'm a repo bot-- nice to meet you!

It has been 30 days since there have been any updates or new comments on this page.  If this issue has been resolved, feel free to disregard the rest of this message.  On the other hand, if you are still waiting on a patch, please:

  + review our [contribution guide](https://github.com/balderdashy/sails/blob/master/CONTRIBUTING.md) to make sure this submission meets our criteria (only _verified bugs_ with documented features, please;  no questions, commentary, or bug reports about undocumented features or unofficial plugins)
  + create a new issue with the latest information, including updated version details with error messages, failing tests, and a link back to [the original issue]().  This allows GitHub to automatically create a back-reference for future visitors arriving from search engines.

Thanks so much for your help!

**Comment by tkrotoff** [comment]
This sailsbot is pretty annoying... This issue should stay open.

**Comment by devinivy** [comment]
@tjwebb how has been your experience with the postgres module using knex?

**Comment by tjwebb** [comment]
@devinivy extremely positive.

**Comment by harish2704** [comment]
Hi all, I have written a simple module which can transform mongo like queries to knex queries.
These are the [tests] ( https://github.com/harish2704/knex-json-query/blob/master/test.js ) passed at the moment.

I am not sure whether it is relevant in this discussion or not. But simply posting here in the hope that, it will be helpful to someone.

**Comment by atiertant** [comment]
@harish2704 we wrote a knex based adapter, you could have a look at it [here](https://github.com/Atlantis-Software/offshore-sql)

**Comment by hunsche** [comment]
Is required have an ORM that manages migrations.

**Comment by connor4312** [comment]
@hunsche check out [umzug](https://github.com/sequelize/umzug) for a nice, framework-agnostic way of handing migrations. Hooking it up to Waterline should take minutes.
